File: reflexuniflow/reflexuniflow/components/focus.py
import reflex as rx
from reflexuniflow.states.focus_state import FocusState


def mode_pill(label: str, value: str, icon: str) -> rx.Component:
    return rx.el.button(
        rx.icon(icon, class_name="h-4 w-4"),
        rx.el.span(label),
        on_click=lambda: FocusState.set_mode(value),
        class_name=rx.cond(
            FocusState.mode == value,
            "flex items-center gap-2 px-5 py-2 rounded-full bg-[#b45309] text-white font-medium text-sm shadow-sm",
            "flex items-center gap-2 px-5 py-2 rounded-full text-[#5c3a1e] font-medium text-sm hover:bg-[#f5e6c8]",
        ),
    )


def circular_progress() -> rx.Component:
    return rx.el.div(
        rx.el.div(
            rx.el.div(
                class_name="absolute inset-0 rounded-full",
                style={
                    "background": f"conic-gradient(#b45309 {FocusState.progress_percent}%, #f5e6c8 0)"
                },
            ),
            rx.el.div(
                rx.el.p(
                    FocusState.display_time,
                    class_name="text-5xl sm:text-6xl font-bold text-[#5c3a1e] font-['IBM_Plex_Mono'] tabular-nums",
                ),
                rx.cond(
                    FocusState.mode == "pomodoro",
                    rx.el.p(
                        rx.cond(
                            FocusState.pomodoro_phase == "work",
                            "FOCUS TIME",
                            "BREAK TIME",
                        ),
                        class_name=rx.cond(
                            FocusState.pomodoro_phase == "work",
                            "text-xs font-bold text-[#b45309] font-['IBM_Plex_Mono'] tracking-widest mt-2",
                            "text-xs font-bold text-[#0d9488] font-['IBM_Plex_Mono'] tracking-widest mt-2",
                        ),
                    ),
                    rx.el.p(
                        "TIMER",
                        class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] tracking-widest mt-2",
                    ),
                ),
                class_name="absolute inset-4 rounded-full bg-[#fbf3e2] flex flex-col items-center justify-center",
            ),
            class_name="relative w-72 h-72 sm:w-80 sm:h-80 rounded-full",
        ),
        class_name="flex items-center justify-center",
    )


def cycle_dot(dot) -> rx.Component:
    return rx.el.div(
        class_name=rx.cond(
            dot["filled"] == "true",
            "h-3 w-3 rounded-full bg-[#b45309]",
            "h-3 w-3 rounded-full bg-[#e8d5a8]",
        )
    )


def controls() -> rx.Component:
    return rx.el.div(
        rx.cond(
            FocusState.running,
            rx.el.button(
                rx.icon("pause", class_name="h-5 w-5"),
                rx.el.span("Pause"),
                on_click=FocusState.pause,
                class_name="flex items-center gap-2 px-6 py-3 rounded-2xl bg-[#b45309] text-white font-semibold hover:bg-[#92400e]",
            ),
            rx.el.button(
                rx.icon("play", class_name="h-5 w-5"),
                rx.el.span("Start"),
                on_click=FocusState.start,
                class_name="flex items-center gap-2 px-6 py-3 rounded-2xl bg-[#b45309] text-white font-semibold hover:bg-[#92400e]",
            ),
        ),
        rx.el.button(
            rx.icon("rotate-ccw", class_name="h-5 w-5"),
            "Reset",
            on_click=FocusState.reset_timer,
            class_name="flex items-center gap-2 px-6 py-3 rounded-2xl border-2 border-[#e8d5a8] text-[#5c3a1e] font-semibold hover:bg-[#f5e6c8]",
        ),
        rx.cond(
            FocusState.mode == "pomodoro",
            rx.el.button(
                rx.icon("skip-forward", class_name="h-5 w-5"),
                "Skip",
                on_click=FocusState.skip_phase,
                class_name="flex items-center gap-2 px-6 py-3 rounded-2xl border-2 border-[#e8d5a8] text-[#5c3a1e] font-semibold hover:bg-[#f5e6c8]",
            ),
            rx.fragment(),
        ),
        class_name="flex flex-wrap items-center justify-center gap-3",
    )


def pomodoro_settings() -> rx.Component:
    return rx.cond(
        FocusState.mode == "pomodoro",
        rx.el.div(
            rx.el.div(
                rx.el.p(
                    "Cycles",
                    class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] uppercase mb-2",
                ),
                rx.el.div(
                    rx.foreach(FocusState.cycle_dots, cycle_dot),
                    class_name="flex items-center gap-2",
                ),
                class_name="text-center",
            ),
            rx.el.div(
                rx.el.label(
                    "Total cycles",
                    class_name="text-xs text-[#8b6f47] font-semibold block mb-1",
                ),
                rx.el.input(
                    type="number",
                    default_value=FocusState.total_cycles.to_string(),
                    on_change=FocusState.set_total_cycles.debounce(300),
                    class_name="w-24 px-3 py-2 rounded-lg border-2 border-[#e8d5a8] bg-[#faf1dd] text-sm text-[#3d2914]",
                ),
            ),
            class_name="flex items-center gap-6 flex-wrap justify-center",
        ),
        rx.el.div(
            rx.el.label(
                "Duration (minutes)",
                class_name="text-xs text-[#8b6f47] font-semibold block mb-1",
            ),
            rx.el.input(
                type="number",
                default_value=FocusState.timer_duration_minutes.to_string(),
                on_change=FocusState.set_timer_duration.debounce(300),
                class_name="w-32 px-3 py-2 rounded-lg border-2 border-[#e8d5a8] bg-[#faf1dd] text-sm text-[#3d2914]",
            ),
            class_name="text-center",
        ),
    )


def stats_row() -> rx.Component:
    return rx.el.div(
        rx.el.div(
            rx.el.p(
                "Study today",
                class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] uppercase",
            ),
            rx.el.p(
                f"{FocusState.study_minutes_today}m",
                class_name="text-2xl font-bold text-[#5c3a1e] font-['Fraunces']",
            ),
            class_name="flex-1 bg-[#fbf3e2] p-4 rounded-2xl border-2 border-[#e8d5a8] text-center",
        ),
        rx.el.div(
            rx.el.p(
                "Pomodoros",
                class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] uppercase",
            ),
            rx.el.p(
                FocusState.pomodoros_completed_today,
                class_name="text-2xl font-bold text-[#5c3a1e] font-['Fraunces']",
            ),
            class_name="flex-1 bg-[#fbf3e2] p-4 rounded-2xl border-2 border-[#e8d5a8] text-center",
        ),
        rx.el.div(
            rx.el.p(
                "Timers done",
                class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] uppercase",
            ),
            rx.el.p(
                FocusState.timers_completed_today,
                class_name="text-2xl font-bold text-[#5c3a1e] font-['Fraunces']",
            ),
            class_name="flex-1 bg-[#fbf3e2] p-4 rounded-2xl border-2 border-[#e8d5a8] text-center",
        ),
        rx.el.div(
            rx.el.p(
                "All time",
                class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] uppercase",
            ),
            rx.el.p(
                FocusState.study_hours_all,
                class_name="text-2xl font-bold text-[#5c3a1e] font-['Fraunces']",
            ),
            class_name="flex-1 bg-[#fbf3e2] p-4 rounded-2xl border-2 border-[#e8d5a8] text-center",
        ),
        class_name="grid grid-cols-2 sm:grid-cols-4 gap-3 mt-6",
    )


def session_row(s) -> rx.Component:
    return rx.el.div(
        rx.el.div(
            rx.cond(
                s["completed"] == "yes",
                rx.icon("circle-check", class_name="h-4 w-4 text-[#0d9488]"),
                rx.icon("circle-dashed", class_name="h-4 w-4 text-[#8b6f47]"),
            ),
            rx.el.span(
                s["mode"],
                class_name="text-sm font-semibold text-[#5c3a1e]",
            ),
            class_name="flex items-center gap-2",
        ),
        rx.el.div(
            rx.el.span(
                s["duration"],
                class_name="text-sm font-bold text-[#b45309] font-['IBM_Plex_Mono']",
            ),
            rx.el.span(
                s["date"],
                class_name="text-xs text-[#8b6f47] font-['IBM_Plex_Mono']",
            ),
            class_name="flex items-center gap-3",
        ),
        class_name="flex items-center justify-between px-3 py-2 bg-[#faf1dd] rounded-xl border border-[#e8d5a8]",
    )


def recent_sessions_section() -> rx.Component:
    return rx.cond(
        FocusState.recent_sessions.length() > 0,
        rx.el.div(
            rx.el.div(
                rx.icon("history", class_name="h-4 w-4 text-[#b45309]"),
                rx.el.span(
                    "Recent sessions",
                    class_name="text-xs font-bold text-[#8b6f47] font-['IBM_Plex_Mono'] uppercase tracking-widest",
                ),
                class_name="flex items-center gap-2 mb-3",
            ),
            rx.el.div(
                rx.foreach(FocusState.recent_sessions, session_row),
                class_name="flex flex-col gap-2",
            ),
            class_name="bg-[#fbf3e2] p-5 rounded-2xl border-2 border-[#e8d5a8] mt-6",
        ),
        rx.fragment(),
    )


def focus_content() -> rx.Component:
    return rx.el.div(
        rx.el.div(
            rx.el.h1(
                "Focus",
                class_name="text-4xl font-bold text-[#5c3a1e] font-['Fraunces']",
            ),
            rx.el.p(
                "Deep work, one session at a time.",
                class_name="text-[#8b6f47] mt-1",
            ),
            class_name="mb-6 text-center",
        ),
        rx.el.div(
            mode_pill("Pomodoro", "pomodoro", "brain"),
            mode_pill("Timer", "timer", "hourglass"),
            class_name="flex items-center justify-center gap-2 mb-8 bg-[#fbf3e2] p-1.5 rounded-full w-fit mx-auto border-2 border-[#e8d5a8]",
        ),
        circular_progress(),
        rx.el.div(controls(), class_name="mt-8"),
        rx.el.div(pomodoro_settings(), class_name="mt-6"),
        stats_row(),
        recent_sessions_section(),
    )


def focus_page() -> rx.Component:
    from reflexuniflow.components.nav import page_shell

    return page_shell(focus_content())

File: UniFlow/uniflow/static/js/views/focus.js
// Focus page: pomodoro/timer mode switch, the countdown dial, and session
// stats. The dial's digits and progress ring are re-painted directly by
// timers.js every 100ms rather than through a full re-render — see that file.
// `rebaseFocus` is called after actions that reset the clock server-side
// while it keeps running (skip), where the next state push won't do it for us.

import { AMBIENT_OPTIONS, currentPrefs, play, setVolume, stop as stopAmbient } from "../ambient.js";
import { h } from "../dom.js";
import { countdown } from "../format.js";
import { icon } from "../icons.js";
import { act } from "../state.js";
import { clock, rebaseFocus } from "../timers.js";
import { eyebrow, tile } from "../ui.js";

function modePill(focus, { label, value, icon: iconName }) {
  return h(
    "button",
    {
      class: ["pill", focus.mode === value && "is-active"],
      onClick: async () => {
        await act("focus/mode", { mode: value });
        rebaseFocus(0);
      },
    },
    icon(iconName, "icon-sm"),
    h("span", {}, label),
  );
}

function dial(focus) {
  const targetDs = focus.target_seconds * 10;
  const percent = targetDs > 0 ? Math.min(100, (clock.focusDs / targetDs) * 100) : 0;
  const isWork = focus.pomodoro_phase === "work";

  return h(
    "div",
    { class: "dial-wrap" },
    h(
      "div",
      { class: "dial" },
      h("div", { class: "dial-ring", id: "focus-ring", style: { "--progress": `${percent}%` } }),
      h(
        "div",
        { class: "dial-inner" },
        h("p", { class: "p dial-time", id: "focus-time" }, countdown(targetDs - clock.focusDs)),
        focus.mode === "pomodoro"
          ? h("p", { class: ["p dial-phase", isWork ? "is-work" : "is-break"] }, isWork ? "FOCUS TIME" : "BREAK TIME")
          : h("p", { class: "p dial-phase" }, "TIMER"),
      ),
    ),
  );
}

function controls(focus) {
  return h(
    "div",
    { class: "row row-wrap", style: { justifyContent: "center", gap: "12px" } },
    focus.running
      ? h(
          "button",
          {
            class: "btn btn-lg btn-primary",
            onClick: () => act("focus/pause", { elapsed_ds: clock.focusDs }),
          },
          icon("pause"),
          "Pause",
        )
      : h(
          "button",
          { class: "btn btn-lg btn-primary", onClick: () => act("focus/start") },
          icon("play"),
          "Start",
        ),
    h(
      "button",
      { class: "btn btn-lg btn-outline", onClick: () => act("focus/reset") },
      icon("rotate-ccw"),
      "Reset",
    ),
    focus.mode === "pomodoro"
      ? h(
          "button",
          {
            class: "btn btn-lg btn-outline",
            onClick: async () => {
              await act("focus/skip", { elapsed_ds: clock.focusDs });
              rebaseFocus(0);
            },
          },
          icon("skip-forward"),
          "Skip",
        )
      : null,
  );
}

function settings(focus) {
  if (focus.mode === "pomodoro") {
    return h(
      "div",
      { class: "row row-wrap", style: { justifyContent: "center", gap: "24px" } },
      h(
        "div",
        { class: "center" },
        h("p", { class: "p eyebrow mb-2" }, "Cycles"),
        h(
          "div",
          { class: "cycle-dots" },
          focus.cycle_dots.map((dot) => h("div", { class: ["dot", dot.filled && "is-filled"] })),
        ),
      ),
      h(
        "div",
        {},
        h("label", { class: "label label-sm", for: "focus-cycles" }, "Total cycles"),
        h("input", {
          class: "input input-sm input-tiny",
          id: "focus-cycles",
          type: "number",
          min: "1",
          value: String(focus.total_cycles),
          onChange: (event) => act("focus/settings", { total_cycles: event.target.value }),
        }),
      ),
    );
  }
  return h(
    "div",
    { class: "row row-wrap", style: { justifyContent: "center", gap: "16px" } },
    h(
      "div",
      { class: "center" },
      h("label", { class: "label label-sm", for: "focus-duration-min" }, "Minutes"),
      h("input", {
        class: "input input-sm input-narrow",
        id: "focus-duration-min",
        type: "number",
        min: "0",
        value: String(focus.timer_duration_minutes),
        // No rebase here: changing the target while idle is already reflected
        // (elapsed is 0 either way), and forcing it while running would zero
        // out an in-progress countdown the server was never told to reset.
        onChange: (event) => act("focus/settings", { timer_duration_minutes: event.target.value }),
      }),
    ),
    h(
      "div",
      { class: "center" },
      h("label", { class: "label label-sm", for: "focus-duration-sec" }, "Seconds"),
      h("input", {
        class: "input input-sm input-narrow",
        id: "focus-duration-sec",
        type: "number",
        min: "0",
        max: "59",
        value: String(focus.timer_duration_seconds),
        onChange: (event) => act("focus/settings", { timer_duration_seconds: event.target.value }),
      }),
    ),
  );
}

// Ambient playback lives entirely client-side (see ambient.js), so clicks
// here update the DOM directly instead of round-tripping through act()/a
// full state re-render.
function ambientControls() {
  const prefs = currentPrefs();
  const pills = [];

  const applyActiveState = (kind) => {
    for (const pill of pills) {
      const isActive = pill.dataset.kind === kind;
      pill.classList.toggle("is-active", isActive);
    }
  };

  const optionPill = (option) => {
    const pillEl = h(
      "button",
      {
        class: ["pill", option.value === prefs.kind && "is-active"],
        type: "button",
        onClick: () => {
          if (option.value === "none") stopAmbient();
          else play(option.value);
          applyActiveState(option.value);
        },
      },
      option.label,
    );
    pillEl.dataset.kind = option.value;
    pills.push(pillEl);
    return pillEl;
  };

  const volumeSlider = h("input", {
    type: "range",
    class: "ambient-volume",
    min: "0",
    max: "100",
    value: String(Math.round(prefs.volume * 100)),
    "aria-label": "Ambient volume",
    onInput: (event) => setVolume(Number(event.target.value) / 100),
  });

  return h(
    "div",
    { class: "card card-lg mt-6" },
    eyebrow("infinity", "Ambient sound"),
    h("div", { class: "switch mb-3", style: { flexWrap: "wrap" } }, AMBIENT_OPTIONS.map(optionPill)),
    h(
      "div",
      { class: "row", style: { gap: "10px", maxWidth: "260px" } },
      icon("volume", "icon-sm"),
      volumeSlider,
    ),
  );
}

function sessionRow(session) {
  return h(
    "div",
    { class: "session-row" },
    h(
      "div",
      { class: "row" },
      icon(session.completed === "yes" ? "circle-check" : "circle-dashed", "icon-sm"),
      h("span", { class: "small", style: { fontWeight: "600", color: "var(--heading)" } }, session.mode),
    ),
    h(
      "div",
      { class: "row", style: { gap: "12px" } },
      h("span", { class: "small mono", style: { fontWeight: "700", color: "var(--accent)" } }, session.duration),
      h("span", { class: "tiny muted mono" }, session.date),
    ),
  );
}

export function focusPage(state) {
  const { focus } = state;

  return h(
    "div",
    {},
    h(
      "div",
      { class: "center mb-6" },
      h("h1", { class: "h1" }, "Focus"),
      h("p", { class: "p muted mt-2" }, "Deep work, one session at a time."),
    ),
    h(
      "div",
      { class: "switch mb-8" },
      modePill(focus, { label: "Pomodoro", value: "pomodoro", icon: "brain" }),
      modePill(focus, { label: "Timer", value: "timer", icon: "hourglass" }),
    ),
    dial(focus),
    h("div", { class: "mt-8" }, controls(focus)),
    h("div", { class: "mt-6" }, settings(focus)),
    ambientControls(),
    h(
      "div",
      { class: "grid grid-2 grid-tight grid-sm-4 mt-6" },
      tile("Study today", `${focus.study_minutes_today}m`),
      tile("Pomodoros", focus.pomodoros_completed_today),
      tile("Timers done", focus.timers_completed_today),
      tile("All time", focus.study_hours_all),
    ),
    focus.recent_sessions.length
      ? h(
          "div",
          { class: "card card-lg mt-6" },
          eyebrow("history", "Recent sessions"),
          h("div", { class: "stack-sm" }, focus.recent_sessions.map(sessionRow)),
        )
      : null,
  );
}
